exporter/newrelicexporter/newrelic.go

```diff
@@ -170,101 +103,119 @@ func (e *exporter) extractInsertKeyFromHeader(ctx context.Context) string {
+	details := newLogMetadata(ctx)
+	details.dataInputCount = ld.LogRecordCount()
+	return e.export(&details, ctx, func() (telemetry.PayloadEntry, error) {
+		var (
+			errs  []error
+			batch telemetry.LogBatch
+		)
+
+		for i := 0; i < ld.ResourceLogs().Len(); i++ {
+			resourceLogs := ld.ResourceLogs().At(i)
+			resource := resourceLogs.Resource()
+
+			for j := 0; j < resourceLogs.InstrumentationLibraryLogs().Len(); j++ {
+				instrumentationLibraryLogs := resourceLogs.InstrumentationLibraryLogs().At(j)
+
+				transformer := newTransformer(resource, instrumentationLibraryLogs.InstrumentationLibrary())
+				for k := 0; k < instrumentationLibraryLogs.Logs().Len(); k++ {
+					log := instrumentationLibraryLogs.Logs().At(k)
+					nrLog, err := transformer.Log(log)
+					if err != nil {
+						e.logger.Error("Transform of log failed.", zap.Error(err))
+						errs = append(errs, err)
+						continue
+					}
+
+					details.dataOutputCount++
+					batch.Logs = append(batch.Logs, nrLog)
+				}
 			}
-			batch.Spans = append(batch.Spans, spans...)
 		}
-	}
-	batches := []telemetry.PayloadEntry{&batch}
-	var req *http.Request
-	var err error
 
-	if insertKey != "" {
-		req, err = e.requestFactory.BuildRequest(batches, telemetry.WithInsertKey(insertKey))
-	} else {
-		req, err = e.requestFactory.BuildRequest(batches)
-	}
-	if err != nil {
-		sentCount = 0
-		e.logger.Error("Failed to build batch", zap.Error(err))
-		return err
-	}
+		return &batch, consumererror.CombineErrors(errs)
+	})
+}
 
-	// Execute the http request and handle the response
-	httpStatusCode, err := e.doRequest(details, req)
-	if err != nil {
-		// We also treat downstream service unavailability as successful for our purposes
-		if httpStatusCode != http.StatusForbidden && httpStatusCode != http.StatusServiceUnavailable {
-			sentCount = 0
+func (e exporter) pushMetricData(ctx context.Context, md pdata.Metrics) (outputErr error) {
+	details := newMetricMetadata(ctx)
+	_, details.dataInputCount = md.MetricAndDataPointCount()
+	return e.export(&details, ctx, func() (telemetry.PayloadEntry, error) {
+		var (
+			errs  []error
+			batch telemetry.MetricBatch
+		)
+
+		rms := md.ResourceMetrics()
+		_, dpCount := md.MetricAndDataPointCount()
+		batch.Metrics = make([]telemetry.Metric, 0, dpCount)
+		for i := 0; i < rms.Len(); i++ {
+			rm := rms.At(i)
+			ilms := rm.InstrumentationLibraryMetrics()
+			for j := 0; j < ilms.Len(); j++ {
+				ilm := ilms.At(j)
+				ms := ilm.Metrics()
+				transform := newTransformer(rm.Resource(), ilm.InstrumentationLibrary())
+				for k := 0; k < ms.Len(); k++ {
+					m := ms.At(k)
+					nrMetrics, err := transform.Metric(m)
+					if err != nil {
+						e.logger.Debug("Transform of metric failed.", zap.Error(err))
+						errs = append(errs, err)
+						continue
+					}
+					details.dataOutputCount += len(nrMetrics)
+					batch.Metrics = append(batch.Metrics, nrMetrics...)
+				}
+			}
 		}
```

Comment:
If you reorganized it as follows, we'd be able to independently test the preparation of the payload from the measurement / sending of the http request (and all the code would be indented one fewer tab):
```
func (e exporter) pushMetricData(ctx context.Context, md pdata.Metrics) (outputErr error) {
	details := newMetricMetadata(ctx)
	_, details.dataInputCount = md.MetricAndDataPointCount()
	bb := func () (telemetry.PayloadEntry, error) {
		return e.buildMetricPayload(details, ctx, md)
	}
	return e.export(&details, ctx, bb);
}

func (e exporter) buildMetricPayload(details exportMetadata, ctx context.Context, md pdata.Metrics) (telemetry.PayloadEntry, error) {
	var (
		errs  []error
		batch telemetry.MetricBatch
	)

	rms := md.ResourceMetrics()
	_, dpCount := md.MetricAndDataPointCount()
	batch.Metrics = make([]telemetry.Metric, 0, dpCount)
	for i := 0; i < rms.Len(); i++ {
		rm := rms.At(i)
		ilms := rm.InstrumentationLibraryMetrics()
		for j := 0; j < ilms.Len(); j++ {
			ilm := ilms.At(j)
			ms := ilm.Metrics()
			transform := newTransformer(rm.Resource(), ilm.InstrumentationLibrary())
			for k := 0; k < ms.Len(); k++ {
				m := ms.At(k)
				nrMetrics, err := transform.Metric(m)
				if err != nil {
					e.logger.Debug("Transform of metric failed.", zap.Error(err))
					errs = append(errs, err)
					continue
				}
				details.dataOutputCount += len(nrMetrics)
				batch.Metrics = append(batch.Metrics, nrMetrics...)
			}
		}
	}

	return &batch, consumererror.CombineErrors(errs)
}
```